src/roles/dullahan.py

```diff
@@ -103,100 +100,88 @@ def on_del_player(evt, cli, var, nick, nickrole, nicktpls, death_triggers):
             evt.params.del_player(cli, target, True, end_game=False, killer_role=nickrole, deadlist=evt.params.deadlist, original=evt.params.original, ismain=False)
             evt.data["pl"] = evt.params.refresh_pl(pl)
 
-@event_listener("rename_player")
-def on_rename(evt, cli, var, prefix, nick):
-    kvp = []
-    for a,b in KILLS.items():
-        if a == prefix:
-            a = nick
-        if b == prefix:
-            b = nick
-        kvp.append((a,b))
-    KILLS.update(kvp)
-    if prefix in KILLS:
-        del KILLS[prefix]
-    kvp = []
-    for a,b in TARGETS.items():
-        nl = set()
-        for n in b:
-            if n == prefix:
-                n = nick
-            nl.add(n)
-        if a == prefix:
-            a = nick
-        kvp.append((a,nl))
-    TARGETS.update(kvp)
-    if prefix in TARGETS:
-        del TARGETS[prefix]
-
 @event_listener("night_acted")
 def on_acted(evt, cli, var, nick, sender):
-    if nick in KILLS:
+    if users._get(nick) in KILLS: # FIXME
         evt.data["acted"] = True
 
+@event_listener("swap_player")
+def on_swap(evt, var, old_user, user):
+    if old_user in KILLS:
+        KILLS[user] = KILLS.pop(old_user)
+    if old_user in TARGETS:
+        TARGETS[user] = TARGETS.pop(old_user)
+
+    for dullahan, target in KILLS.items():
+        if target is old_user:
+            KILLS[dullahan] = user
+
+    for dullahan, targets in TARGETS.items():
+        if old_user in targets:
+            targets.remove(old_user)
+            targets.add(user)
+
 @event_listener("get_special")
 def on_get_special(evt, cli, var):
     evt.data["special"].update(var.ROLES["dullahan"])
 
 @event_listener("transition_day", priority=2)
 def on_transition_day(evt, cli, var):
-    for k, d in list(KILLS.items()):
-        evt.data["victims"].append(d)
-        evt.data["onlybywolves"].discard(d)
-        evt.data["killers"][d].append(k)
-        del KILLS[k]
+    while KILLS:
+        k, d = KILLS.popitem()
+        evt.data["victims"].append(d.nick)
+        evt.data["onlybywolves"].discard(d.nick)
+        evt.data["killers"][d.nick].append(k.nick)
 
 @event_listener("exchange_roles")
 def on_exchange(evt, cli, var, actor, nick, actor_role, nick_role):
-    if actor in KILLS:
-        del KILLS[actor]
-    if nick in KILLS:
-        del KILLS[nick]
-    if actor_role == "dullahan" and nick_role != "dullahan" and actor in TARGETS:
-        TARGETS[nick] = TARGETS[actor] - {nick}
-        del TARGETS[actor]
-    elif nick_role == "dullahan" and actor_role != "dullahan" and nick in TARGETS:
-        TARGETS[actor] = TARGETS[nick] - {actor}
-        del TARGETS[nick]
+    for k in set(KILLS):
```

Comment:
Except that's not actually possible - `KILLS` doesn't store nicks, it stores users. Therefore, we need to check for that. When that event is updated to use the new API, we won't need to do that anymore, but for now, we do. And nope, shoe-horning in a few `users._get` calls won't work - that function is O(n), and you'd probably call it several times.